### PR TITLE
ACR 8/19 Roadmap Updates

### PR DESCRIPTION
**Pull request by SteveLasker**

### CODE REVIEW

**Comment by SteveLasker** [comment]
@sajayantony

File: docs/acr-roadmap.md [hunk 1]
```diff
@@ -1,117 +1,159 @@
...
-![](./media/storage-donut.png)
+Starting fall '19, [Azure Security Center][asc] will start offering registry scanning and runtime protection. As with any new offering, it will take time for us to dial in the experiences.
 
+We believe customers should have a choice for their security scanning solutions and will continue to provide extensibility, just as users can choose the security scanning of their users machines and their servers.
 
-This was just confusing, and we joked about making it an overflowing cupcake. We settled on simply showing just the facts:
+### Limit Endpoint Access /VNet Support
 
-![](./media/storage-non-cupcake.png)
+Customers have asked for limitations on their registry, based on the IP, not just authentication. As a shared registry API, this does present some challenges that we're incrementally addressing.
```
Comment (sajayantony) [inline docs/acr-roadmap.md]: Don't know if limitations can be taken as positive. I'm guessing the idea here is that customers want more granular access control in terms of who and how the registry is accessed.

File: docs/acr-roadmap.md [hunk 2]
```diff
@@ -1,117 +1,159 @@
...
 
-### Image Promotion ###
-We're starting to see a pattern where teams use multiple registries that align to their environments. The Dev team has a registry they develop, build, test within. The entire development team has full access to push and pull. The production environment is locked down to a subset of users and service accounts. It has only the images that have passed quality testing. It may be the registry that gets geo-replicated, where the dev registry is limited to the region(s) where developers work/deploy to. 
+In [March of 2019 we added preview support for VNet and Firewall rules](https://azure.microsoft.com/en-us/blog/azure-container-registry-virtual-network-and-firewall-rules-preview-support/). This allows customers to limit their acr endpoint to requests from within the defined Azure VNet, or any whitelisted ip ranges. ACR VNet support is supported with AKS and VMs hosting container images. The container host infrastructure must be placed within the customers VNet as the host must be able to pull the image, from the registry that's in the VNet.
```
Comment (sajayantony) [inline docs/acr-roadmap.md]: limit or restrict?

File: docs/acr-roadmap.md [hunk 3]
```diff
@@ -1,117 +1,159 @@
...
 
-We currently have VNet and Firewall rules available in private preview. Please contact us at acrsup@microsoft.com for being added to the VNet Private Preview. 
+With OCI Artifacts in place, we are working with the [Helm Community](https://github.com/helm/) to add [Helm Registry support](https://v3.helm.sh/docs/topics/registries/), allowing customers to use `helm registry push|pull|` across all OCI compliant Registries. As Helm 3 Registry support releases, and the community moves to Helm 3, our plan is to deprecate the Helm 2 acr support for, what we believe is a more native Helm 3 experience. We will continue to support `az acr helm` support for a period of transition to Helm 3. All Helm charts pushed to ACR with `az acr helm` will be fully compatible with Helm 3 registry support. Customers using ACR Helm should be comfortable with a smooth transition to Helm 3 and continue with `az acr helm` in the meantime.
```
Comment (sajayantony) [inline docs/acr-roadmap.md]: Helm 3 and 2 might be breaking and this way we would actually even deprecate az acr helm if helm client is as rich as docker.

File: docs/acr-roadmap.md [hunk 4]
```diff
@@ -1,117 +1,159 @@
...
+When issuing `docker pull` a manifest is returned, listing the layers required. If the local host already has the layers, no additional requests are made. However, the registry must still respond with the manifest. ACR introduced a caching layer to cache & return manifest without having to hit storage. We should mention that an alternative approach would be to use [Web Hooks][webhooks] to know when an image has been updated, rather than simply asking the registry: *"do you have anything new, do you have anything new, do you have anything new..."*.
+We've also been working with a few high profile customers to scale thousands of nodes, each pulling large images. We recently scaled 1,000 nodes in 10 minutes with 11 terabytes being deployed. This performance work has been moved into the Premium SKU to enable high scale customers a solution to their larger needs. We continue to focus on concurrent throughput and individual pull performance scenarios.
```
Comment (sajayantony) [inline docs/acr-roadmap.md]: Should we reword high profile to something like customers that require large scale?

File: docs/acr-roadmap.md (hunk 3 above)
Comment (SteveLasker) [reply docs/acr-roadmap.md]: Not sure what exact change you're suggesting. Ping me today and I can make another PR.

File: docs/acr-roadmap.md (hunk 4 above)
Comment (SteveLasker) [reply docs/acr-roadmap.md]: Changed to early adopters

File: docs/acr-roadmap.md (hunk 3 above)
Comment (iowillhoit) [reply docs/acr-roadmap.md:118]: > All Helm charts pushed to ACR with `az acr helm` will be fully compatible with Helm 3 registry support. 

@SteveLasker @sajayantony Could you please clarify what this means? 

We _were_ pushing charts to ACR with Helm 3 OCI but discovered that Spinnaker [does not yet have OCI support](https://github.com/spinnaker/spinnaker/issues/5474). We are switching to pushing charts with `az acr helm` to get our Spinnaker pipelines working. We are trying to decide if we should "dual push" charts with both `az acr helm push` (helm 2) and `helm chart push` (helm 3 OCI) to be ready for when Spinnaker supports OCI. 

This comment makes me think that charts pushed with `az acr helm push` will someday be accessible with helm 3 OCI commands and also show up in the Portal in the ACR UI. Thanks!

**Comment by SteveLasker** [comment]
Calling @bacongobbler, bridgetkromhout, squillace to comment on Helm integration with Spinaker as well as Terraform.

For ACR migration of Helm 2 charts to Helm 3 OCI Artifact support, this turned out to be more difficult than planned. It was my hope we could pull this off, transparently. Because Helm 2 charts are stored outside an OCI registry, we had no clear place to put them. If I pushed a `hello-wordpress` chart to my demo42.azurecr.io registry, where would you expect it to show up under OCI Artifacts? `demo42.azurecr.io/hello-wordpress` might be used by something else. `demo42.azurecr.io/helm-charts/hello-wordpress` might be used by something else. 
We've discussed some reserved repos using ~, _ or some other characters: `demo42.azurecr.io/~helm-charts/hello-wordpress`, but the distribution spec doesn't support this. We could add it, but that's another process.

So, we're hoping to see the Helm community promote Helm registry support to the main release and customers migrate their charts. We won't deprecate helm 2 support until customers have moved to helm 3. Our goal is to help customers, not frustrate them :)

Steve

File: docs/acr-roadmap.md (hunk 3 above)
Comment (sajayantony) [reply docs/acr-roadmap.md:118]: @ iowillhoit - Helm 2 charts aren't stored the same way as OCI charts are so we can't auto migrate the content over. You would have to migrate and repush them as Helm3 charts. We don't have a timeline for deprecation and just want to make sure the the path going forward is clear for helm2 and helm3 authors.

**Comment by bacongobbler** [comment]
> So, we're hoping to see the Helm community promote Helm registry support to the main release and customers migrate their charts.

There are several big ticket items that are preventing us from marking OCI support as stable; the most notable of which being chart provenance support. I don't foresee this work being complete any time soon unless someone from the community steps up to complete the work.

For context: our team is focused on maintaining Helm, consulting the community on what work they'd like to implement for the next Helm release, reviewing PRs from the community, and ensuring that we are releasing new updates in a timely manner. We will step in to fix issues related to security, but we're not making any investments to implement new features at this time. We're leaving that to the community so we can focus on other projects.

Hope that helps provide some context on our roles as Helm maintainers.

**Comment by iowillhoit** [comment]
Thanks for the clarification @SteveLasker and @sajayantony! The potential name conflicts with a migration was going to be my next question. We will continue to use Helm 2 for the time being and make the call internally if we want to "dual push" or script a migration later when Spinnaker supports OCI. Thanks again 👍

**Pull request merged**